pkg/cloudprovider/cloudprovider.go

```diff
@@ -106,6 +106,11 @@ func (c *CloudProvider) Create(ctx context.Context, nodeClaim *karpv1.NodeClaim)
 	if nodeClassReady.IsUnknown() {
 		return nil, cloudprovider.NewCreateError(fmt.Errorf("resolving NodeClass readiness, NodeClass is in Ready=Unknown, %s", nodeClassReady.Message), "NodeClassReadinessUnknown", "NodeClass is in Ready=Unknown")
 	}
+	// Prevent NodeClaim creation in the case where spec.role changes but nodeclass status has not updated with appropriate instance profile
+	instanceProfileCondition := nodeClass.StatusConditions().Get(v1.ConditionTypeInstanceProfileReady)
```

Comment:
I believe we should be able to use the observed generation on the `Ready` condition rather than the `InstanceProfile` condition. This solves this general class of issues. Each time we update and condition, we [recompute the root (Ready) condition](https://github.com/awslabs/operatorpkg/blob/57066b748e19d092c9700fda03c7a941e760dfd4/status/condition_set.go#L232-L261). If any of the root's dependent conditions don't have a matching observed generation or if they're false or unknown, the root condition goes to false or unknown.

Reply:
Other than the change to use the root condition though, this logic looks good.